In [1]:
import pandas as pd
import numpy as np
from ast import literal_eval

# Load in Grouped Data File

In [2]:
generic = lambda x: literal_eval(x)

conv = {'url_id_path': generic,
        'seconds_spent_path': generic}

df = pd.read_csv('DATA/Data/testGrouped.csv', converters=conv)

# Load in Test Data File
### Used to test the validity and correctness of the below methods

In [3]:
# load test data
testDF = pd.read_csv('DATA/Data/testVis.csv', converters=conv)

# Method to get the number of visits which visit an unscraped URL

In [4]:
def get_number_of_visits_to_unscraped_urls():
    count_visits_to_unscraped_urls = 0
    for index, row in df.iterrows():
        found_unscraped_url = False
        for url_id in row.url_id_path:
            # If there is a new visit which visits an unscraped URL, increase counter
            if url_id > 1559 and not found_unscraped_url:
                count_visits_to_unscraped_urls += 1
                found_unscraped_url = True
        if index % 100000 == 0: # Update on progress as can take quite a few minutes.
            print(f"{index} rows complete...")
    print("DONE!")
    return count_visits_to_unscraped_urls

# The % of visits which visit an unscraped URL
11% of visits visit an unscraped URL.

# Method to get number of visits less than 20 seconds and path length of those visits

In [27]:
def get_number_of_visits_less_20_seconds_and_their_path_lengths():
    visits_less_20_seconds = 0
    path_lengths = []
    times_spent = []
    for index, row in df.iterrows():
        timeSpent = sum(row.seconds_spent_path)
        if timeSpent < 20: # Change here for less than or greater than 20
            visits_less_20_seconds += 1
            path_lengths.append(len(row.url_id_path)) # Keep track of path lengths for sub 20 seconds visits
            times_spent.append(timeSpent)
        if index % 100000 == 0:
            print(f"{index} rows complete...")

    print("DONE!")
    return [times_spent, visits_less_20_seconds, path_lengths]

In [ ]:
returned = get_number_of_visits_less_20_seconds_and_their_path_lengths()
times_20 = returned[0]
visits_less_20 = returned[1]
path_lengths_under_20 = returned[2]

In [30]:
print(visits_less_20/len(df.index))

0.3356965426419714


In [31]:
print(sum(path_lengths_under_20)/len(path_lengths_under_20))

1.1697623223123246


In [32]:
print(sum(times_20)/len(times_20))

1.9427805207294973


# Information about visits that are less than 20 seconds in duration

## The % of visits that are less than 20 seconds in duration
33.5% of visits are less than 20 seconds in duration.

## The average path lengths in visits under 20 seconds in duration
1.1697623223123246 is the average path length.

## Average time spent in visits under 20 seconds in duration
1.9427805207294973 is the average time spent.

In [ ]:
visits_greater_20, path_lengths = get_number_of_visits_less_20_seconds_and_their_path_lengths()

In [20]:
print(visits_greater_20/len(df.index))

0.6643034573580285


In [21]:
print(sum(path_lengths)/len(path_lengths))

7.4080256085265805


In [ ]:
times_spent = get_number_of_visits_less_20_seconds_and_their_path_lengths()

In [22]:
print(sum(times_spent)/len(times_spent))

482.03758769414577


# Information about visits over or equal to 20 seconds

## The % of visits over or equal to 20 seconds
66.4% of visits are over or equal to 20 seconds

## The average path length in visits over or equal to 20 seconds
7.4 Pages visited in the path on average

## The average time spent by visitors with visits over or equal to 20 seconds
482 seconds or over 8 minutes!
